### PR TITLE
[DEPRECATION]Deprecate send action (RFC #335)

### PR DESCRIPTION
**Pull request by cibernox**
Ready for review.

### CODE REVIEW

File: packages/ember-views/lib/mixins/action_support.js [hunk 1]
```diff
@@ -116,6 +116,10 @@ export default Mixin.create({
       `Attempted to call .sendAction() with the action '${action}' on the destroyed object '${this}'.`,
       !this.isDestroying && !this.isDestroyed
     );
+    deprecate('Component#sendAction is deprecated. Please use closure actions instead.', false, {
```
Comment (rwjblue) [inline packages/ember-views/lib/mixins/action_support.js]: We should include `this` somewhere in the deprecation message (to help track down where the deprecation is coming from).

**Review by rwjblue — COMMENTED** [review]
Looking good! I left an inline comment RE: including more identifiable info in the deprecation message, and I'd like to see the `sendAction` support code be wrapped for svelting as part of this PR.

The svelte wrapping stuff is pretty new, but _should_ hopefully be easy to add. I'll dig up an example...

**Comment by rwjblue** [comment]
See https://github.com/emberjs/ember.js/pull/16745 for an example of the work needed to svelte...

File: packages/ember-views/lib/mixins/text_support.js [hunk 2]
```diff
@@ -297,20 +297,32 @@ export default Mixin.create(TargetActionSupport, {
...
-  let action = get(view, `attrs.${eventName}`) || get(view, eventName);
+  let actionName = get(view, `attrs.${eventName}`) || get(view, eventName);
   let value = get(view, 'value');
 
-  view.sendAction(eventName, value);
+  if (typeof actionName === 'string') {
+    deprecate(
+      `Passing actions to components as strings (like {{input ${eventName}="${actionName}"}}) is deprecated. Please use closure actions instead ({{input ${eventName}=(action "${actionName}")}})`,
+      false,
+      { id: 'ember-input.send-action', until: '4.0.0' }
```
Comment (cibernox) [inline packages/ember-views/lib/mixins/text_support.js]: Note to self: Use the same id than in the `sendAction` deprecation even if the text is different, as in it's essence it's the same feature.

**Comment by cibernox** [comment]
@rwjblue I've implemented the AST deprecation and tested it. I've also wrapped the `sendAction` helper for svelte removal.

Points of attention: 
1. Should the AST deprecation to have the same ID as the runtime deprecation?
2. In the file of constants for svelte, what's the version number I should use?

**Comment by cibernox** [comment]
PR for the deprecation entry: https://github.com/ember-learn/deprecation-app/pull/146

File: packages/ember-views/lib/mixins/text_support.js [hunk 3]
```diff
@@ -297,20 +297,36 @@ export default Mixin.create(TargetActionSupport, {
...
 // In principle, this shouldn't be necessary, but the legacy
 // sendAction semantics for TextField are different from
 // the component semantics so this method normalizes them.
 function sendAction(eventName, view, event) {
-  let action = get(view, `attrs.${eventName}`) || get(view, eventName);
+  let actionName = get(view, `attrs.${eventName}`) || get(view, eventName);
   let value = get(view, 'value');
 
-  view.sendAction(eventName, value);
+  if (typeof actionName === 'string') {
```
[the remark this answers is not in the document]
Comment (rwjblue) [reply packages/ember-views/lib/mixins/text_support.js]: Also, how does this guard support?

```hbs
{{input action='some-generic-action'}}
```

File: packages/ember-template-compiler/lib/plugins/deprecate-send-action.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  'escape-press',
+  'focus-in',
+  'focus-out',
+  'key-press',
+  'key-up',
+  'key-down',
+];
+
+export default function deprecateRender(env: ASTPluginEnvironment): ASTPlugin | undefined {
+  let { moduleName } = env.meta;
```
[the remark this answers is not in the document]
Comment (rwjblue) [reply packages/ember-template-compiler/lib/plugins/deprecate-send-action.ts]: file size reduction

[the remark this answers is not in the document]
Comment (cibernox) [reply packages/ember-template-compiler/lib/plugins/deprecate-send-action.ts]: I thought this AST transforms were not shipped as templates are compiled to JSON.

**Comment by cibernox** [comment]
@rwjblue can you rebuild the browserstack job? I think it's a false negative.

**Comment by rwjblue** [comment]
Done!

**Comment by cibernox** [comment]
Green!

**Review by rwjblue — APPROVED** [review]

**Comment by rwjblue** [comment]
The deprecation RFC has been merged now, so this is good to go.

Thanks @cibernox!

**Comment by cibernox** [comment]
@rwjblue the docs also need a push 😄   
https://github.com/ember-learn/deprecation-app/pull/146

**Pull request merged**